docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,498 @@
+  - Otherwise, let `M0, ..., Mn`be the default bounds for `<T0 extends A0,
+      ..., Tn extends An>`
+
+If `[M0, ..., Mn]` do not satisfy the bounds `<T0 extends B0, ..., Tn extends
+Bn>` then it is an inference failure error.
+
+Otherwise, the full solution is `[T0 -> M0, ..., Tn -> Mn]`.
+
+### Downwards generic instantiation resolution
+
+Downwards resolution is the process by which the return type of a generic method
+(or constructor, etc) is matched against a type expectation context from an
+invocation of the method to find a (possibly partial) solution for the missing
+type arguments
+
+`[T0 -> P0, ..., Tn -> Pn]` is a partial solution for a set of type variables
+`<T0 extends B0, ..., Tn extends Bn>` under constraint set `Cp` given a type
+expectation of `R` with respect to a return type `Q` (in which the `Ti` may be
+free) where the `Pi` are type schemas (potentially just `?` if unresolved)/
+
+If `R <: Q [T0, ..., Tn] -> C` does not hold, then each `Pi` is `?` and `Cp` is
+    empty
+
+Otherwise:
+  - `R <: Q [T0, ..., Tn] -> C` and `Cp` is `C`
+  - If `C` does not constrain `Ti` then `Pi` is `?`
+  - If `C` partially constrains `Ti`
+    - If `C` is over constrained, then it is an inference failure error
+    - Otherwise `Pi` is the constraint solution for `Ti` with respect to `C`
+  - If `C` fully constrains `Ti`, then
+    - Let `Ai` be `Bi[R0/T0, ..., ?/Ti, ..., ?/Tn]`
+    - If `C + Ti <: Ai` is over constrained, it is an inference failure error.
+    - Otherwise, `Ti` is fixed to be `Pi`, where `Pi` is the constraint solution
+      for `Ti` with respect to `C + Ti <: Ai`.
+
+### Upwards generic instantiation resolution
+
+Upwards resolution is the process by which the parameter types of a generic
+method (or constructor, etc) are matched against the actual argument types from
+an invocation of a method to find a solution for the missing type arguments that
+have not been fixed by downwards resolution.
+
+`[T0 -> M0, ..., Tn -> Mn]` is the upwards solution for an invocation of a
+generic method of type `<T0 extends B0, ..., Tn extends Bn>(P0, ..., Pk) -> Q`
+given actual argument types `R0, ..., Rk`, a partial solution `[T0 -> P0, ...,
+Tn -> Pn]` and a partial constraint set `Cp`:
+  - If `Ri <: Pi [T0, ..., Tn] -> Ci`
+  - And the full constraint resolution of `Cp + C0 + ... + Cn` for `<T0 extends
+B0, ..., Tn extends Bn>` given the initial partial solution `[T0 -> P0, ..., Tn
+-> Pn]` is `[T0 -> M0, ..., Tn -> Mn]`
+
+### Discussion
+
+The incorporation of the type bounds information is asymmetric and procedural:
+it iterates through the bounds in order (`Bi[R0/T0, ..., ?/Ti, ..., ?/Tn]`).  Is
+there a better formulation of this that is symmetric but still allows some
+propagation?
+
+### Inference rules
+
+- The expression `e as T` is inferred as `m as T` of type `T` in context `K`:
+  - If `e` is inferred as `m` in an empty context
+- The expression `x = e` is inferred as `x = m` of type `T` in context `K`:
```

Comment:
I confess that in my implementation work so far I have been largely ignoring this section and following my intuition because I had a difficult time following the style in which these rules are written.  This morning I figured out the reason I've been confused: because the language used here makes the data flow and the recursion implicit, and mixes inputs and outputs.  For example, on this particular line, it was not clear to me for a long time that `x = e` and `K` are inputs to the inference algorithm, whereas `x = m` and `T` are outputs.  Nor the fact that `T` and `m` are defined to be the outputs of the recursive call to type inference in the line below.

Also I found it confusing that the exposition treats `e` and `m` as two distinct expressions, whereas in my intuitive understanding of the algorithm (as well as in both of our implementations of it) they are represented by the same data structure at different times (`e` is the expression before it is inferred and `m` is the expression afterwards).

Consider rephrasing this section in a way that makes the recursive nature of the algorithm more explicit, makes the data flow more procedural, and uses `e'` rather than `m` to emphasize that it is conceptually the same expression tree, but with type inference information added.  For example, maybe these two lines could be rewritten something like this:

- To infer `x = e` in context `K`:
  - Let `M` be the type of `x`.
  - Apply type inference to `e` in context `M`, producing `e'` with type `T`.
  - Then the result of inference is `x = e'` with type `T`.

(Note: I'm considering trying to rework this section on my own, just to make sure I understand it.  If I do, I'll share the resulting text with you and you can decide whether you want to use any of it.)